### ISSUE TITLE
[UX] Provide a report page that lists all configured paths where files are being saved.

### ISSUE THREAD

**Issue by klonos** [backdrop/backdrop-issues#4585]
@olafgrabienski in https://github.com/backdrop/backdrop-issues/issues/4583#issuecomment-683414706

> ... In text format configuration (e.g. `admin/config/content/formats/filtered_html`) you can "Enable file uploads", set an Upload directory and more. It's quite hidden, and I often forget this option when building new sites.

We should have a report page, where site builders/admins can get an overview of all the various directories where files are being saved for text formats, media types, file/image fields etc.

---

Advocate: @klonos

**Comment by klonos** [comment]
@stpaultim I thought we should get this one out of the way as well: https://github.com/backdrop/backdrop/pull/3634

Here's what I have so far:
![image](https://user-images.githubusercontent.com/2423362/119393264-64343400-bcd9-11eb-9bd1-5c9dd8bc6a1c.png)

...and with some more directories + private dir configured:
![image](https://user-images.githubusercontent.com/2423362/119392978-14ee0380-bcd9-11eb-92c7-2a8296b8507e.png)

**Comment by klonos** [comment]
...feedback I'm looking for:

- Does "Storage" make sense for the middle column? The alternative was "Scheme", which I don't find particularly "human-friendly".
- Are the "Public files"/"Private files" cells readable/scannable enough, or too similar? The alternative was to use plain "public"/"private". Here's how that would look: 
![image](https://user-images.githubusercontent.com/2423362/119395843-c2164b00-bcdc-11eb-8767-7251a410b324.png)
- Does the "This is the default/root directory..." text make sense? Suggestions?

Things I need to work on:
- "pretty" human-names for the entity names for the field usage list items (especially things like "node").

**Comment by indigoxela** [comment]
I did some super sloppy testing in the sandbox. The first thing that caught my eye:

```
Notice: Undefined index: file_upload in file_upload_dir_list() (line 699 of /var/lib/tugboat/core/modules/file/file.admin.inc).
```

**Comment by klonos** [comment]
@indigoxela thanks for testing this. I can do some defensive coding, but curious to know how you managed to trigger that error.

**Comment by indigoxela** [comment]
> curious to know how you managed to trigger that error

Nothing special, I turned on error display and went to https://pr3634-05xu7lvykqhrckpa6sgi8q3sg7agcb9r.tugboat.qa/admin/reports/file_upload_dir_list

**Comment by klonos** [comment]
Huh 👀

**Comment by klonos** [comment]
Thanks @indigoxela 🙏🏼 ...fixed.

**Comment by stpaultim** [comment]
I'm just looking at this in the sandbox. Looks helpful. 

I have not had time to process what I'm seeing yet, as I'm a bit overwhelmed with other things right now, but have some quick feedback. 

Would "Access" work better than "Storage"? It might not be technically correct, but it feels like public/private is an access issue, although different than other access issues in Backdrop. I think that "Storage" is fine, since anyone reading this page will (I think) understand the difference between public/private regardless of the label. 

I notice that the "temp" directory is not listed. Is there a reason for that?

**Comment by klonos** [comment]
> I have not had time to process what I'm seeing yet, as I'm a bit overwhelmed with other things right now...

Right ...I meant to mention @olafgrabienski, since he was the one that requested this; sorry I mentioned you instead @stpaultim but thanks for taking the time to test anyway 😅 

In theory, there may be more "[stream wrappers](https://www.drupal.org/node/555118#s-streamwrapper-notation-for-all-files)" than just public/private, so I'm not sure "Storage" would cover all these cases.

> I notice that the "temp" directory is not listed. Is there a reason for that?

The temp directory is only used as (as the name suggests) a temporary storage until the file is uploaded/saved. This folder doesn't actually hold any media assets in the CMS. I could include it in the list, but then we'd need to come up with a better name for the report, which I'm currently calling "File upload directories" 🤷🏼

**Comment by indigoxela** [comment]
> Would "Access" work better than "Storage"?

IMO "storage" is correct. (Not a native speaker, though.)

I took a look at the file system settings page (help text):

```
 A local file system path where public files will be stored.
 An existing local file system path for storing private files.
```

So, "storage" seems consistent to me - we're using the same term there.

> In theory, there may be more "stream wrappers" than just public/private

But if files get stored there (wherever it is), it's also a "storage", right?

**Comment by olafgrabienski** [comment]
> I meant to mention @olafgrabienski, since he was the one that requested this

@klonos Thanks for working on this issue! I'm also busy with other things and can't test right now. I wanted however mention that I didn't "request" anything in regard of this issue, see the discussion starting at https://github.com/backdrop/backdrop-issues/issues/4583#issuecomment-683414706.

**Comment by klonos** [comment]
> But if files get stored there (wherever it is), it's also a "storage", right?

Yup, that's where I was getting at 👍🏼 

>... I didn't "request" anything in regard of this issue...

Oops! Sorry if I interpreted that this way. Still a nice feature though, right? 😅 

I'll wait for people to have more time to test this and get more feedback/ideas/thoughts then. Thanks everyone for having a look 🙏🏼

**Comment by klonos** [comment]
Note to self: from the dev meeting today, @quicksketch suggested making sure that contrib can use this feature, to add any locations in the filesystem that are configured to store files uploaded. It could be something similar to https://github.com/indigoxela/backdrop/blob/1.x/core/modules/image/image.api.php#L160

**Comment by BWPanda** [comment]
I like the idea. Some thoughts:

- Public/private are very similar, probably not fixable though. But I think we should capitalise the terms here.
- The last column's text seems unnecessarily verbose. Can we simplify that?

**Comment by klonos** [comment]
> But I think we should capitalise the terms here.

Easy 👍🏼 ...do others agree with this though? (so that I don't go back and forth in making changes to the PR)

> Can we simplify that?

Sure. Suggestions? ...please keep in mind that my main goal with that text was to provide inks to the places where each respective path can be edited.

**Comment by BWPanda** [comment]
I don't think we need to add 'files' to public/private. That just adds more similar text that makes it harder to distinguish.

As for re-wording, how about just:
- Documents field, Page content type
- Documents field, Tags vocabulary
- Filtered HTML text format
- etc.

**Comment by klonos** [comment]
Thanks @BWPanda 👍🏼 ...please note that I also output the field type. So unless you have a better idea, I was thinking this:

- Private docs **file** field, Page content type
- Avatar **image** field, User account user account type
- Filtered HTML text format **images**

**Comment by klonos** [comment]
...I've updated the PR. Here's what we have now:

![image](https://user-images.githubusercontent.com/2423362/120047024-36722680-c01c-11eb-9a25-5efe75763e21.png)

**Comment by klonos** [comment]
...last tweak (not in the PR yet): does everyone think that the indicators for default dirs is fine as a bullet point, as in the screenshot in the comment above, or like so:

![image](https://user-images.githubusercontent.com/2423362/120050384-e5ffc680-c025-11eb-89d9-e92eb9d3d3d4.png)

...or does anyone have any better idea re how to indicate those in the list?

**Comment by BWPanda** [comment]
Ooh, I like that last one better!

**Comment by klonos** [comment]
I think I like it better too. I'll wait for some more feedback, and will then update the PR accordingly.

**Comment by olafgrabienski** [comment]
I also like the version *without* bullet point better. In this case, I'd prefer a normal font-weight.

**Comment by stpaultim** [comment]
> I also like the version without bullet point better. In this case, I'd prefer a normal font-weight.

I agree. 

This seems like a nice report, but I'm not sure that I'd ever use it. I'm not against having it in core, but I'm also not completely convinced of the need for it. Will defer to the opinions of others.

**Comment by BWPanda** [comment]
There's a conflict in the PR, and while I could fix it myself in @klonos' absence, it seems that the PR is changing files outside of the scope of this issue... So I think we need to wait for @klonos to come and clear this up. Marking as NW until then.
